runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -5870,128 +5906,124 @@ TR::Register *J9::Power::TreeEvaluator::VMmonexitEvaluator(TR::Node *node, TR::C
-      //    andc    threadReg, monitorReg, threadReg
-      // #endif
-      //    cmp    cr0, 0, metaReg, threadReg
-      //    bne    callLabel                   // inflated or flc bit set
-      //    addi   monitorReg, monitorReg, -LOCK_INC_DEC_VALUE
-      //    st     [objReg, monitor offset], monitorReg
-      // doneLabel:
-      // callReturnLabel:
-      // === OUT OF LINE ===
-      // callLabel:
-      //    bl     jitMonitorExit
-      //    b      callReturnLabel
-
-      TR::LabelSymbol *decLabel, *doneLabel;
-      metaReg = cg->getMethodMetaDataRegister();
+      lockSize = 4;
+      loadOpCode = TR::InstOpCode::lwz;
+      storeOpCode = TR::InstOpCode::stw;
+      reservedLoadOpCode = TR::InstOpCode::lwarx;
+      conditionalStoreOpCode = TR::InstOpCode::stwcx_r;
+      }
 
-      decLabel = generateLabelSymbol(cg);
+   /*
+    * Read only locks do not support either the Reserved bit nor the Learning bit.
+    * They are disabled until the code path is updated.
+    */
+   if (true || !ppcSupportsReadMonitors || !node->isReadMonitor())
+      {
+      /*
+       * JIT unlocking fast path checks:
+       * 1. Check if lockword matches TID and the lower 8 bits are clear. If so, clear lockword to unlock the object.
+       * 2. Check if TID matches, Learning bit is clear, and Inflated bit is clear. If not, call VM.
+       * 3. Check if RC field is at least 1. If so, decrement RC.
+       * 4. All fast path checks failed so go to VM.
+       *
+       * Check 1 catches Flat-Locked (non-nested, FLC bit clear).
+       * Check 2 catches New-PreLearning, New-AutoReserve, Learning-Unlocked, Learning-Locked, Flat-Unlocked and Inflated and sends them to the VM.
+       * Check 3 catches Flat-Locked (nested case) and Reserved-Locked.
+       * Check 4 catches Reserved-Unlocked and Flat-Locked (non-nested case, FLC bit set) and sends it to the VM.
+       *
+       *
+       *    ld/lwz monitorReg, lwOffset(baseReg)
+       *    li     tempReg, 0
+       *    cmpl   cr0, monitorReg, metaReg
+       *    bne    decrementCheckLabel
+       *    lwsync
+       *    st     tempReg, lwOffset(baseReg)
+       *    b      doneLabel
+       *
+       * decrementCheckLabel:
+       * #if 64BIT
+       *    li     tempReg, LOCK_NON_PRIMITIVE_EXIT_IGNORE_MASK
+       *    andc   tempReg, monitorReg, tempReg
+       * #else 32BIT
+       *    rlwinm tempReg, monitorReg, 0, ~LOCK_NON_PRIMITIVE_EXIT_IGNORE_MASK
+       * #endif
+       *    cmpl   cr0, tempReg, metaReg
+       *    bne    callLabel
+       *    andi.  tempReg, monitorReg, OBJECT_HEADER_LOCK_RECURSION_MASK
+       *    beq    callLabel
+       *    addi   monitorReg, monitorReg, -LOCK_INC_DEC_VALUE
+       *    st     monitorReg, lwOffset(baseReg)
+       *
+       * doneLabel:
+       * callReturnLabel:
+       * === OUT OF LINE ===
+       * callLabel:
+       *    bl    jitMonitorExit
+       *    b     callReturnLabel
+       */
```

Comment:
Performance implications of code layout ... have you looked into it?  i.e. more frequently travelled code should be arranged to fall through, avoiding taken branches which hardware will take anywhere 2-4 cycles to turn around, even assuming that it is perfectly predicted.

Reply:
I've tried to minimize the number of taken branches on the fast path. Most of the taken branches are done when a check says to go to the VM to handle the current state. There's also a branch that selects between the nested and non-nested case that can't be avoided. It currently falls through to the non-nested case since that's what the original code did.